### PR TITLE
mgr/dashboard: Fixed name clash when hostname similar to another

### PR DESCRIPTION
**Pull request by wornet-mwo**
Fixes issue #49769, where displayed statistics showed irritating values when hostname are similar like "node1" and "node10".

Signed-Off-By: Michael Wodniok <wodniok@wor.net>

Here are some screenshots before and after. The Node in question has no OSDs. For details see comments of this PR.
Before:
rz2b-cn1 uncorrected
After:
rz2b-cn1 corrected

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary, not necessary
- [ ] Includes tests for new functionality or reproducer for bug, not necessary

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: monitoring/grafana/dashboards/host-details.json [hunk 1]
```diff
@@ -37,7 +37,7 @@
   "gnetId": null,
   "graphTooltip": 0,
   "id": null,
-  "iteration": 1557386759572,
+  "iteration": 1615564911000,
```
Comment (avanthakkar) [inline monitoring/grafana/dashboards/host-details.json:40]: @wornet-mwo Thank you for your contribution. Is this change really required?

Comment (wornet-mwo) [reply monitoring/grafana/dashboards/host-details.json:40]: AFAIK this causes grafana to notify the client (browser) to refresh the dashboard. Maybe it isn't in fact but it's usually increased if a dashboard is edited in the frontend.

**Comment by avanthakkar** [comment]
@wornet-mwo Can you attach before/after screenshots? Just to verify the changes resolved the issue.

**Comment by wornet-mwo** [comment]
@avanthakkar Okay, sorry, did not have time for this before. The case disappered because of an hardware defect resulting in removal of the node in question. Now i could reproduce this using a vm with no OSDs. Here we go:
We have named our nodes rz2b-cn<X> where X is an increasing number. rz2b-cn1 is a vm with no OSDs, 10, 11, 12 are other nodes which break the graphs. Following shows the situation with current mainlined graph:
rz2b-cn1 uncorrected
With applied corrections this looks correctly for rz2b-cn1 (no OSDs as expected):
rz2b-cn1 corrected
And the other nodes still look correctly, here rz2b-cn12 (the host has 6 OSDs):
rz2b-cn12 corrected

**Comment by avanthakkar** [comment]
> @avanthakkar Okay, sorry, did not have time for this before. The case disappered because of an hardware defect resulting in removal of the node in question. Now i could reproduce this using a vm with no OSDs. Here we go:
> We have named our nodes rz2b-cn where X is an increasing number. rz2b-cn1 is a vm with no OSDs, 10, 11, 12 are other nodes which break the graphs. Following shows the situation with current mainlined graph:
> rz2b-cn1 uncorrected
> With applied corrections this looks correctly for rz2b-cn1 (no OSDs as expected):
> rz2b-cn1 corrected
> And the other nodes still look correctly, here rz2b-cn12 (the host has 6 OSDs):
> rz2b-cn12 corrected

LGTM, thank you @wornet-mwo !

**Review by avanthakkar — APPROVED** [review]

**Review by aaSharma14 — APPROVED** [review]
Thanks @wornet-mwo

**Review by p-se — CHANGES_REQUESTED** [review]
I tested these changes on my development environment (first) and didn't see any graphs on the host details dashboard anymore. On the development environment, the host name does not contain a FQDN with dots, only a colon with the port: `localhost:9100`.

The regular expression (`(\..*)`) you've used seems to introduce a regression by breaking  simple non-FQDN host names.

The regular expression suffix `([.:].*)` (as compared to `(\..*)`) for matching the instance label seemed to fix both on my tests, the issue you've found and the regression introduced by these changes.

**Comment by wornet-mwo** [comment]
> I tested these changes on my development environment (first) and didn't see any graphs on the host details dashboard anymore. On the development environment, the host name does not contain a FQDN with dots, only a colon with the port: `localhost:9100`.
> 
> The regular expression (`(\..*)`) you've used seems to introduce a regression by breaking simple non-FQDN host names.

That's a different issue. A hostname can not contain colons according to RFC 952/1123. Somehow the colon seems to get into prometheus - so we have to decide how to solve the issue. @avanthakkar, @aaSharma14: Which way do you prefer? A: Loosen the regex / B: Fix the wrong hostname (surely more work)

If B: How can i reproduce the behaviour, could you point me to the right place?

**Comment by p-se** [comment]
@wornet-mwo 

The [`Hostname`](https://github.com/ceph/ceph/blob/master/monitoring/grafana/dashboards/host-details.json#L1165) variable used in Grafana is determined from the [value of the `instance` label of a metric (the metric seems to be `node_scrape_collector_success` mostly)](https://github.com/ceph/ceph/blob/master/monitoring/grafana/dashboards/host-details.json#L1169). On my development environment, the instance labels' value is `localhost:9100` but on your environment it is something like `host1.domain.com:9100`. Grafana template variables are defined [per dashboard](https://github.com/ceph/ceph/blob/master/monitoring/grafana/dashboards/host-details.json#L1159-L1179). The instance value is then parsed [using a regex that takes everything up to a dot or colon](https://github.com/ceph/ceph/blob/master/monitoring/grafana/dashboards/host-details.json#L1171). This is what ends up being in the `Hostname` field of the dashboards and is used for the [`$ceph_hosts` variable](https://github.com/ceph/ceph/blob/master/monitoring/grafana/dashboards/host-details.json#L1167) inside the dashboard templates. 

That means that whatever host names are used, they will only be displayed without FQDN in Grafana dashboards, as the `$ceph_hosts` variable will never contain a hostname with FQDN. Note that the [regex used to parse the value of the instance label](https://github.com/ceph/ceph/blob/master/monitoring/grafana/dashboards/host-details.json#L1171) says "take whatever but a dot or a colon", so it is natural to use a regex that contains the parsed value and adds the rule "is followed by a dot or colon" when `$ceph_hosts` is used to match a value of an instance label. 

Basically Grafana takes a value of a label from Prometheus to introduce a variable which is then used to filter Prometheus' results dynamically (mostly on a per-host basis). I can think of two things to "fix the hostnames":
1. Changing the value of the `instance` label in Prometheus for more convenience or clarity in Grafana. But that would (ast most) just be that, an improvement for more convenience or clarity, but not what I'd consider to be a real fix for this issue. IMHO, it could also be confusing and it might be better to introduce a new label named `hostname` for that. I don't even know if that would really fix the issue without introducing new ones, so it would probably involve a lot of testing, too.
2. Changing how the `$ceph_hosts` variable is parsed to "take everything up to a colon", which would result in less friction when querying Prometheus using `$ceph_hosts` but also in having FQDN hostnames in Grafana dashboards, should they be used in a cluster. So basically a change of behavior and a fix for this issue at the same time.

But what I do consider to be just a fix for this issue is to adapt the regex to match how Grafana parses the `instance` label as proposed in my [previous comment](https://github.com/ceph/ceph/pull/40072#pullrequestreview-624553063).

Please also share your point of view @avanthakkar and @aaSharma14.

**Comment by avanthakkar** [comment]
> @wornet-mwo
> 
> The [`Hostname`](https://github.com/ceph/ceph/blob/master/monitoring/grafana/dashboards/host-details.json#L1165) variable used in Grafana is determined from the [value of the `instance` label of a metric (the metric seems to be `node_scrape_collector_success` mostly)](https://github.com/ceph/ceph/blob/master/monitoring/grafana/dashboards/host-details.json#L1169). On my development environment, the instance labels' value is `localhost:9100` but on your environment it is something like `host1.domain.com:9100`. Grafana template variables are defined [per dashboard](https://github.com/ceph/ceph/blob/master/monitoring/grafana/dashboards/host-details.json#L1159-L1179). The instance value is then parsed [using a regex that takes everything up to a dot or colon](https://github.com/ceph/ceph/blob/master/monitoring/grafana/dashboards/host-details.json#L1171). This is what ends up being in the `Hostname` field of the dashboards and is used for the [`$ceph_hosts` variable](https://github.com/ceph/ceph/blob/master/monitoring/grafana/dashboards/host-details.json#L1167) inside the dashboard templates.
> 
> That means that whatever host names are used, they will only be displayed without FQDN in Grafana dashboards, as the `$ceph_hosts` variable will never contain a hostname with FQDN. Note that the [regex used to parse the value of the instance label](https://github.com/ceph/ceph/blob/master/monitoring/grafana/dashboards/host-details.json#L1171) says "take whatever but a dot or a colon", so it is natural to use a regex that contains the parsed value and adds the rule "is followed by a dot or colon" when `$ceph_hosts` is used to match a value of an instance label.
> 
> Basically Grafana takes a value of a label from Prometheus to introduce a variable which is then used to filter Prometheus' results dynamically (mostly on a per-host basis). I can think of two things to "fix the hostnames":
> 
>     1. Changing the value of the `instance` label in Prometheus for more convenience or clarity in Grafana. But that would (ast most) just be that, an improvement for more convenience or clarity, but not what I'd consider to be a real fix for this issue. IMHO, it could also be confusing and it might be better to introduce a new label named `hostname` for that. I don't even know if that would really fix the issue without introducing new ones, so it would probably involve a lot of testing, too.
> 
>     2. Changing how the `$ceph_hosts` variable is parsed to "take everything up to a colon", which would result in less friction when querying Prometheus using `$ceph_hosts` but also in having FQDN hostnames in Grafana dashboards, should they be used in a cluster. So basically a change of behavior and a fix for this issue at the same time.
> 
> 
> But what I do consider to be just a fix for this issue is to adapt the regex to match how Grafana parses the `instance` label as proposed in my [previous comment](https://github.com/ceph/ceph/pull/40072#pullrequestreview-624553063).
> 
> Please also share your point of view @avanthakkar and @aaSharma14.

I don't have any strong opinion here, but I have my +1 for your  [previous comment](https://github.com/ceph/ceph/pull/40072#pullrequestreview-624553063).

**Review by p-se — CHANGES_REQUESTED** [review]
We still need to address the regression one way or the other. Preferably by simply fixing the regular expression in this PR. I hope https://github.com/ceph/ceph/pull/40072#issuecomment-816530338 explains reasonably well why I think this should be done.

**Comment by wornet-mwo** [comment]
@p-se I've researched a bit about the difference of hostnames and instances. In my case, the instance seems to be just luckily the hostname without port. So your solution is mostly right. I've adapted it (putting just `[.:]` would not solve the clash), please check it again.

**Review by p-se — APPROVED** [review]
Thanks!

**Comment by p-se** [comment]
@wornet-mwo would you mind merging those two commits into (the first) one?

**Comment by avanthakkar** [comment]
jenkins test dashboard

**Review by avanthakkar — APPROVED** [review]

**Comment by wornet-mwo** [comment]
> @wornet-mwo would you mind merging those two commits into (the first) one?

I can do, but i do not know how (except opening a different PR) - do you have a hint which git command is useful for that? My quick search did not bring up a prominent result.

**Comment by aaSharma14** [comment]
> > @wornet-mwo would you mind merging those two commits into (the first) one?
> 
> I can do, but i do not know how (except opening a different PR) - do you have a hint which git command is useful for that? My quick search did not bring up a prominent result.

`git rebase -i HEAD~2` would help here. After running this command you will get (in the terminal) the two commits with `pick` added to the commit title. Keep the first one as it is and change the second pick to `fixup`. That would do the trick.

**Review by aaSharma14 — APPROVED** [review]

**Comment by epuertat** [comment]
@wornet-mwo Exactly as how @aaSharma14 [pointed out](https://www.git-tower.com/learn/git/faq/git-squash/).

**Comment by p-se** [comment]
> > > @wornet-mwo would you mind merging those two commits into (the first) one?
> > 
> > 
> > I can do, but i do not know how (except opening a different PR) - do you have a hint which git command is useful for that? My quick search did not bring up a prominent result.
> 
> `git rebase -i HEAD~2` would help here. After running this command you will get (in the terminal) the two commits with `pick` added to the commit title. Keep the first one as it is and change the second pick to `fixup`. That would do the trick.

@wornet-mwo note that, once you have merged those two commits as described by @aaSharma14, you will not be able to simply `git push` these changes to your branch as you're used to, but you will need to force-push these changes (`git push --force`) so that the history you have in your branch is replaced (the last two commits are replaced by the one that contains changes of both of them). While this should normally never be done on a repository where people collaborate, this is an operation that is safe to do in your own branch (assuming you're the only one working on that branch). The pull request will automatically reflect these changes, when being pushed to the branch this PR is created from, which is exactly what is going to happen if you just issue `git push --force`.

**Comment by wornet-mwo** [comment]
Okay, done. I was aware rebase is capable reordering, removing commits - i've learnt it's also able to squash commits. Thank you for the hint.

**Comment by epuertat** [comment]
jenkins test make check

**Comment by epuertat** [comment]
Thanks @wornet-mwo ! I'll merge this as soon as it passes the failing CI.
